dvc/scm/git/backend/dulwich.py

```diff
@@ -353,20 +353,20 @@ def iter_remote_refs(self, url: str, base: Optional[str] = None):
         try:
             _remote, location = get_remote_repo(self.repo, url)
             client, path = get_transport_and_path(location)
+            if base:
+                yield from (
+                    os.fsdecode(ref)
+                    for ref in client.get_refs(path)
+                    if ref.startswith(os.fsencode(base))
+                )
+            else:
+                yield from (os.fsdecode(ref) for ref in client.get_refs(path))
+
```

Comment:
Why is this code block moved?

`client.get_refs()` failing is not always going to be due to an invalid URL. `get_transport_and_path` does validation on whether or not the specified git remote exists, or whether or not an explicit URL is a valid git URL.

I get that you are trying to validate whether or not the URL is reachable before the experiment is run, but I don't think that is necessary at all. And even if you want to validate the URL by doing an `scm.iter_refs` call, you still don't need to move this code block to do it.

(and if `client.get_refs()` fails, it would be better to pass the actual exception for why it failed up the chain, instead of just assuming it was due to an invalid URL.)

Reply:
> I get that you are trying to validate whether or not the URL is reachable before the experiment is run, but I don't think that is necessary at all. 

Because the first push will happen after one iteration, stopping at that point will leave a dirty workspace.

Reply:
Right, but that's not any different than any other failure that may occur during an experiment run, or during `dvc repro`.

The user's pipeline code itself could fail, or the `dvc push` could fail due to any number of reasons, whether it's because of a network error, or because the remote config is actually invalid. (We don't validate being able to connect to a DVC remote, all we do is validate that a default remote exists in the DVC config).

All of these cases will result in a "dirty" workspace, which is fine.

Reply:
In some cases, training 1 epoch would cost hours or even days of time. And it will be really frustrating if got an error from DVC.

Reply:
My point is that we are not checking for other things that can fail in DVC, after the epoch has been generated.

Just as an example, we do not verify whether or not it is possible to actually push the cache data to the default DVC remote. If I have mis-configured my DVC remote using either a URL that doesn't exist, or with the wrong authentication credentials, we will fail after we have generated a commit and try to auto-push it.

If we are going to do this kind of strict validation on the git remote, then we should also be doing it on the DVC remote. But in my opinion, both of these checks would be excessive.

---

I also still do not think we should be failing at all in the event that the automatic push fails. We should wrap any exceptions that occur when we try to push, and log them if it fails, but continue running the pipeline. If the user has misconfigured the git URL, they can just do a manual `dvc exp push <correct_git_url> ...` after execution has completed, rather than needing to re-run an entire experiment.

But this is maybe a question for @dberenbaum 

(And yes, this manual push isn't possible in the CML use case, but in the CML use case the key differences are that the user has probably already tested their code locally and that the upstream git URL or remote should be coming from the CI provider itself (i.e. from the github/gitlab environment))

Reply:
It looks the changes are outdated, so it seems like some of the specifics have changed, and I'll focus on the general user experience.

I think there was a previous question from @karajan1001 about whether to validate the url before running the experiment. This validation seems useful and I see little downside from a user perspective. Since dvc knows the experiment will be pushed, it would be helpful to avoid issues before spending time to run the experiment. I also think it makes sense to throw an error if validation fails since auto pushing is specifically for cml and similar workflows, where leaving a dirty workspace isn't helpful.

It would actually seem useful to have some basic validation of both the git and dvc remotes. As @pmrowla mentioned, in cml the `DVC_EXP_GIT_REMOTE` is unlikely to be a problem since it's being set automatically. However, the dvc remote is configured by the user, and it will be problematic if the auto push fails in cml.

Feel free to push back with reasons to avoid this validation or with questions of how strictly to validate.

Reply:
> I think there was a previous question from @karajan1001 about whether to validate the url before running the experiment. This validation seems useful and I see little downside from a user perspective.

There are different possible levels of "validation". IMO validating that a `.git/config` remote named `origin` exists or that the string `git@github.com/my/repo.git` is a valid git URL but `s3://some_non_git_url` is not valid is fine. But I don't think we have to bother checking that `git@github.com/my/repo.git` is reachable before trying to push to it.

This is the exact same level of @karajan1001 's current validation for the DVC remote. We check that a default DVC remote exists in the repo config, but we do not do anything to verify that the remote URL is reachable, or that we have the required list/write/move/delete permissions to actually push to it.

IMO this is fine. But trying to connect to each remote before we actually run anything is overkill, and in the event that it fails we can handle it at that time.

> Since dvc knows the experiment will be pushed, it would be helpful to avoid issues before spending time to run the experiment. I also think it makes sense to throw an error if validation fails since auto pushing is specifically for cml and similar workflows, where leaving a dirty workspace isn't helpful.

I don't think that an automatic push errors in general should be considered a failure state for the entire workflow.  A push failure is an error that the user can recover from by just re-trying `dvc exp push` again after their experiment has completed.

So in the event that github or aws goes down, or that a gdrive remote gets API throttled, or any other kind of possible network error occurs (including a remote being unreachable due to a user misconfiguration), we should just log that "automatic push failed with error: ..." and continue running the pipeline.

> However, the dvc remote is configured by the user, and it will be problematic if the auto push fails in cml.

In this case, users should be testing their remote configurations themselves locally before they git commit things into their `.dvc/config`.

And in the event that they misconfigure it, we should just warn them, and continue running the experiment (for the same reasons I already mentioned)

Reply:
At the end of the day I guess this is a product decision, and if we want to get into creating a test file and `dvc push`ing and then removing it, obviously we can do that. But I think it's overkill and an unnecessary extra step.

Reply:
> There are different possible levels of "validation". IMO validating that a `.git/config` remote named `origin` exists or that the string `git@github.com/my/repo.git` is a valid git URL but `s3://some_non_git_url` is not valid is fine. But I don't think we have to bother opening a test connection to `git@github.com/my/repo.git` before trying to push to it.

How can we check that it's a valid git url? I agree that limiting the validation makes sense since it might not make sense to fail for random timeouts that have nothing to do with the remote configuration. Ideally, it seems like a retry might be helpful, but that's out of scope.

> I don't think that an automatic push errors in general should be considered a failure state for the entire workflow. A push failure is an error that the user can recover from by just re-trying `dvc exp push` again after their experiment has completed.

I think you mentioned before how this isn't helpful in the cml context since you won't be able to access the runner and debug. Are you thinking that, if there's something like a temporary network issue, then the next checkpoint might succeed, so the workflow should continue? Or are you thinking of other scenarios like remote executors where it will be possible to debug?

If there is a temporary network error for a single checkpoint push in the middle of the workflow, failing could be bad for the user since they might not be watching a long-running job and might return later to unexpectedly find that the job did not run to completion. That's why I agree that limiting validation makes sense, and I also agree that auto push errors should warn rather than interrupt the entire job.

However, in the initial validation, failing fast seems like a good user experience since they can debug quickly and resubmit their job. For the initial validation, I think it makes sense to do what we can to check if everything is configured correctly and working as expected and fail if not.

Since this feature is for @iterative/cml, it would be good to get their feedback on this.

Reply:
> How can we check that it's a valid git url?

All of the git backends we use have calls that can do this check (to verify that either a string remote name like `origin` exists, or that a URL conforms to what git expects a git URL to look like)

Basically we would just have to make this a standalone scm call:

https://github.com/iterative/dvc/blob/fd76e0991f6dd47e554b874c2efb8c280dee96c0/dvc/scm/git/backend/dulwich.py#L353-L359

> I think you mentioned before how this isn't helpful in the cml context since you won't be able to access the runner and debug. Are you thinking that, if there's something like a temporary network issue, then the next checkpoint might succeed, so the workflow should continue? Or are you thinking of other scenarios like remote executors where it will be possible to debug?

Depending on the issue, yes it's entirely possible that after one exp commit fails to be pushed in CML, the subsequent push could succeed. But really I was just thinking of scenarios where users set up automatic pushing in general, and not necessarily in CI (this becomes relevant for local exp runs in the event that support for this gets added into dvclive)